Fixes #8 feature request: Add Devise Gem with Configurable

vanamraghu opened the pull request:
1. Updated the code base to have only rails api 
2. Only Devise gem is added with below customization
3. Two controllers registrations & sessions controller is added along with enabling the confirmable (This would send or prompt for confirming the mail address)
4. One bug thats existing in devise for sending fake session to the request headers is provided under concerns (rack_session_fix.rb)
5. cors.rb is uncommented to ensur cross origin is enabled so that external application is provided access
6. In user.rb model, enabled the confirmable 
7. In routes.rb overridden the paths of sessions and registrations
8. Created migration file to enable confirmable
9. Uncommented in schema.rb fields related to confirmable
10. Added in gem file, jsonapi-serializer this is needed to convert or serialize ruby object to JSON
11. User Serializer is added and only id, email and created_at is provided as part of user object
12. Registrations and sessions controller is updated with response_to and response_with so that json is returned (as its rails only api, we are not making use of views and only json response is returned with right status code and message)
13. JSON response messages are modified to have below structure
### Response

> { code: 200, message: 'Signed up successfully', data: { id: 5, email: 'xyz@mail.com', created_at: '2024-08-24T08:00:00.7898"}

Resolves #8

### Description
Summary: Addition & configuration of devise gem
Dependencies: Devise, jsonapi-serializer

### Type of change

- [ ] Feature


### How to test this PR?
Once configuration is completed, below api's can be verified

1. Signup, use below api for signingup/registration, below payload shall be used for sending request and response is displayed below

![CleanShot 2024-08-24 at 10 53 13](https://github.com/user-attachments/assets/eb2f4c66-c340-46c0-bbc0-f62792bb7f4d)


2. Run the mailcatcher so that configured mail address in devise.rb should receive confirmation mail as shown below

**- Run mailcatcher using this command on terminal:** `mailcatcher`

It should run like below:
```
Starting MailCatcher v0.10.0
==> smtp://127.0.0.1:1025
==> http://127.0.0.1:1080
```

![CleanShot 2024-08-23 at 23 04 40](https://github.com/user-attachments/assets/8b371a7b-7a34-45fa-851d-572b80d9beb3)

3. Confirm the mail by clicking on the link

![CleanShot 2024-08-23 at 23 05 56](https://github.com/user-attachments/assets/bd9e7ca9-a68e-43bc-93e9-26dffc904440)

4. Once link is clicked, users account would be marked as confirmed and we are good to login 
5. Use below api for login/signin, (expected response should be successful login as provided in the sessions controller)

![CleanShot 2024-08-24 at 10 54 17](https://github.com/user-attachments/assets/2d87717f-ca63-41b6-bf6c-f25556e6e648)

6. Verify for registration by providing same email id, then we would get error as shown below 

![CleanShot 2024-08-24 at 10 54 43](https://github.com/user-attachments/assets/8252ed78-ad1c-4cf9-83aa-effd6df58729)

vanamraghu committed:
Below are the changes:
1. Commented out oxr.update_rates as there is issue while running the server
2. Devise gem is provided in the gem file and Gemfile.lock is updated with the latest changes
3. Devise gem views are added primarily for sign in, sign up & also changing passwords
4. development.rb is updated with the settings for devise gem
5. Devise is created with user model & corresponding migrations and schema is added
6. Routes file is updated with root pointing to home controller
7. Home controller and index view is created

rt4914 commented on app/controllers/home_controller.rb:
```diff
@@ -0,0 +1,4 @@
+class HomeController < ApplicationController
+  def index; end
+end
+
```
Remove this extra blank space.

Also why do we need this controller?
Our backend will have apis only so we don;t need this UI part in rails.

vanamraghu replied:
Completed and removed devise views too, only rails api

rt4914 reviewed (changes requested):
@vanamraghu There was some setup work pending before we can actually work on this PR.

Basically the project needs to be in api only mode.

I have created the PR you can approve, merge and delete the branch and then update your code accordingly.

Basically there should be no UI work in rails.
https://github.com/rt4914/KlinicCon-Backend/pull/29

vanamraghu committed:
Merge branch 'main' of https://github.com/rt4914/KlinicCon-Backend into 8-add-devise-gem-with-configurable

# Conflicts:
#	Gemfile
#	Gemfile.lock
#	app/assets/stylesheets/application.css
#	app/views/layouts/application.html.erb
#	config/initializers/money.rb

vanamraghu committed:
Below are the changes:
1. Merged changes from master

vanamraghu committed:
Below are the changes:
1. Merged changes from master

vanamraghu committed:
Below are the changes:
1. Devise gem file addition
2. routes are configured with devise
3. DB migration table and also locale & initializer with devise configuration

vanamraghu commented:
> @vanamraghu There was some setup work pending before we can actually work on this PR.
> 
> Basically the project needs to be in api only mode.
> 
> I have created the PR you can approve, merge and delete the branch and then update your code accordingly.
> 
> Basically there should be no UI work in rails. #29

@rt4914 
Yes changes are done, branch has only api changes not UI

rt4914 commented on app/models/user.rb:
```diff
@@ -0,0 +1,6 @@
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
```
We need to add `confirmable` too which will also affect the database schema.

vanamraghu replied:
Added confirmable in the model

rt4914 commented on config/locales/devise.en.yml:
```diff
@@ -0,0 +1,65 @@
+# Additional translations at https://github.com/heartcombo/devise/wiki/I18n
```
For now we can keep this file but not sure if we will actually use it.

rt4914 commented on db/migrate/20240822013600_devise_create_users.rb:
```diff
@@ -0,0 +1,44 @@
+      # t.string   :current_sign_in_ip
+      # t.string   :last_sign_in_ip
+
+      ## Confirmable
```
We need confirmable too.

vanamraghu replied:
Added as part of a separate migration

rt4914 reviewed (changes requested):
You will need to add two devise controllers too something like `registrations_controller` and `sessions_controller` and maybe some more controllers for `confirmation` related task.

Also if you see the original video that I attached https://www.youtube.com/watch?v=Wq9i77BUBlQ

Until and unless we don't make the API call from postman we cannot sat that we have configured the Devise gem for APIs correctly.


Also update your PR description.

vanamraghu committed:
Below are the changes:
1. Updated the code base to have only rails api
2. Only Devise gem is added with below customization
3. Two controllers registrations & sessions controller is added along with enabling the confirmable (This would send or prompt for confirming the mail address)
4. One bug thats existing in devise for sending fake session to the request headers is provided under concerns (rack_session_fix.rb)
5. cors.rb is uncommented to ensur cross origin is enabled so that external application is provided access
6. In user.rb model, enabled the confirmable
7. In routes.rb overridden the paths of sessions and registrations
8. Created migration file to enable confirmable
9. Uncommented in schema.rb fields related to confirmable
10. Added in gem file, jsonapi-serializer this is needed to convert or serialize ruby object to JSON
11. User Serializer is added and only id, email and created_at is provided as part of user object
12. Registrations and sessions controller is updated with response_to and response_with so that json is returned (as its rails only api, we are not making use of views and only json response is returned with right status code and message)

vanamraghu commented:
> You will need to add two devise controllers too something like `registrations_controller` and `sessions_controller` and maybe some more controllers for `confirmation` related task.
> 
> Also if you see the original video that I attached https://www.youtube.com/watch?v=Wq9i77BUBlQ
> 
> Until and unless we don't make the API call from postman we cannot sat that we have configured the Devise gem for APIs correctly.
> 
> Also update your PR description.

@rt4914 Updated the PR description and also about testing

rt4914 commented on app/controllers/users/sessions_controller.rb:
```diff
@@ -0,0 +1,29 @@
+    if current_user
+      render json: {
+        status: 200,
+        message: "logged out successfully"
```
Write this as `Logged out successfully`

basically first character capital

vanamraghu replied:
Changed to Logged

rt4914 commented on Gemfile:
```diff
@@ -49,3 +49,8 @@ group :development do
+gem 'rack-cors'
+
+gem 'devise', '~> 4.9'
+gem 'jsonapi-serializer'
```
EOF (End of File) is missing, add one blank line at the end.

vanamraghu replied:
Fixed

rt4914 reviewed (changes requested):
Really great work @vanamraghu 

One important change:
Current json sturcture is 
```
{
status {
  code ...
  message ...
},
data {
}
```


This should be changed to 

```
{
  code ...
  message ...
  data {
   ....
  }
}
```

We don't need status and also we should keep data as sibling to code and message.

vanamraghu commented:
> Really great work @vanamraghu
> 
> One important change: Current json sturcture is
> 
> ```
> {
> status {
>   code ...
>   message ...
> },
> data {
> }
> ```
> 
> This should be changed to
> 
> ```
> {
>   code ...
>   message ...
>   data {
>    ....
>   }
> }
> ```
> 
> We don't need status and also we should keep data as sibling to code and message.

@rt4914 
Fixed as per the new changes in json structure

vanamraghu committed:
Below are the changes:
1. Added blank line @eof (as per review comments)
2. Modified the structure of json as per new changes
3. Modified the message (Logged out successfully)

vanamraghu committed:
Below are the changes:
1. Added blank line @eof (as per review comments)

rt4914 reviewed (approved):
LGTM, Thanks.
